[Executorch][Llama] Make dynamic shape based export selectable

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #4100

Most backends like mps, qnn, coreml dont support dynamic shape. Particularly
for mps, query to .item(), data dependent dynamic shape, results in lowering
error. This error is likely fixable in the delegate however not clear if the
delegate is meant to support dynamic shape or not.

Thus making dynamic shape specific change selectable for now.

Differential Revision: [D59236591](https://our.internmc.facebook.com/intern/diff/D59236591/)

kimishpatel committed:
[Executorch][Llama] Make dynamic shape based export selectable

Most backends like mps, qnn, coreml dont support dynamic shape. Particularly
for mps, query to .item(), data dependent dynamic shape, results in lowering
error. This error is likely fixable in the delegate however not clear if the
delegate is meant to support dynamic shape or not.

Thus making dynamic shape specific change selectable for now.

Differential Revision: [D59236591](https://our.internmc.facebook.com/intern/diff/D59236591/)

[ghstack-poisoned]

cccclai commented on examples/models/llama2/llama_transformer.py:
```diff
@@ -209,23 +212,31 @@ def update(
-        narrowed_k.copy_(k_val)
-        narrowed_v.copy_(v_val)
-        return self.k_cache, self.v_cache
+        if self.enable_dynamic_shape:
```
Let's add a TODO as it's really messy to insert the logic inside the model here.

cccclai reviewed (approved):
Thanks for the quick fix and unblock the other backends.

kimishpatel committed:
Update on "[Executorch][Llama] Make dynamic shape based export selectable"


Most backends like mps, qnn, coreml dont support dynamic shape. Particularly
for mps, query to .item(), data dependent dynamic shape, results in lowering
error. This error is likely fixable in the delegate however not clear if the
delegate is meant to support dynamic shape or not.

Thus making dynamic shape specific change selectable for now.

Differential Revision: [D59236591](https://our.internmc.facebook.com/intern/diff/D59236591/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch][Llama] Make dynamic shape based export selectable"


Most backends like mps, qnn, coreml dont support dynamic shape. Particularly
for mps, query to .item(), data dependent dynamic shape, results in lowering
error. This error is likely fixable in the delegate however not clear if the
delegate is meant to support dynamic shape or not.

Thus making dynamic shape specific change selectable for now.

Differential Revision: [D59236591](https://our.internmc.facebook.com/intern/diff/D59236591/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch][Llama] Make dynamic shape based export selectable"


Most backends like mps, qnn, coreml dont support dynamic shape. Particularly
for mps, query to .item(), data dependent dynamic shape, results in lowering
error. This error is likely fixable in the delegate however not clear if the
delegate is meant to support dynamic shape or not.

Thus making dynamic shape specific change selectable for now.

Differential Revision: [D59236591](https://our.internmc.facebook.com/intern/diff/D59236591/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch][Llama] Make dynamic shape based export selectable"


Most backends like mps, qnn, coreml dont support dynamic shape. Particularly
for mps, query to .item(), data dependent dynamic shape, results in lowering
error. This error is likely fixable in the delegate however not clear if the
delegate is meant to support dynamic shape or not.

Thus making dynamic shape specific change selectable for now.

Differential Revision: [D59236591](https://our.internmc.facebook.com/intern/diff/D59236591/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch][Llama] Make dynamic shape based export selectable"


Most backends like mps, qnn, coreml dont support dynamic shape. Particularly
for mps, query to .item(), data dependent dynamic shape, results in lowering
error. This error is likely fixable in the delegate however not clear if the
delegate is meant to support dynamic shape or not.

Thus making dynamic shape specific change selectable for now.

Differential Revision: [D59236591](https://our.internmc.facebook.com/intern/diff/D59236591/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch][Llama] Make dynamic shape based export selectable"


Most backends like mps, qnn, coreml dont support dynamic shape. Particularly
for mps, query to .item(), data dependent dynamic shape, results in lowering
error. This error is likely fixable in the delegate however not clear if the
delegate is meant to support dynamic shape or not.

Thus making dynamic shape specific change selectable for now.

Differential Revision: [D59236591](https://our.internmc.facebook.com/intern/diff/D59236591/)

[ghstack-poisoned]
